pkg/apis/sources/v1beta1/ping_conversion.go

```
@@ -18,17 +18,130 @@ package v1beta1
+		// Default to a wrapped message.
+		return json.Marshal(message{Body: body})
+	}
+	return json.Marshal(objmap)
+}
+
 // ConvertTo implements apis.Convertible
-func (source *PingSource) ConvertTo(ctx context.Context, sink apis.Convertible) error {
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", sink)
+// Converts source from v1beta1.PingSource into a higher version.
+func (source *PingSource) ConvertTo(ctx context.Context, obj apis.Convertible) error {
+	switch sink := obj.(type) {
+	case *v1beta2.PingSource:
+		sink.ObjectMeta = source.ObjectMeta
+		sink.Status = v1beta2.PingSourceStatus{
+			SourceStatus: source.Status.SourceStatus,
+		}
+
+		// deep copy annotations to avoid mutation on source.ObjectMeta.Annotations
+		annotations := make(map[string]string)
+		for key, value := range source.GetAnnotations() {
+			annotations[key] = value
+		}
+
+		// try to unmarshal v1beta2.PingSource.Spec from V1B2SpecAnnotationKey
+		v1beta2Spec, ok := annotations[V1B2SpecAnnotationKey]
+		if ok {
+			if err := json.Unmarshal([]byte(v1beta2Spec), &sink.Spec); err != nil {
+				ok = false
+			}
+			// we don't need this annotation in a v1beta2.PingSource object
+			delete(annotations, V1B2SpecAnnotationKey)
+		}
+
+		// cannot unmarshal, do a normal conversion
+		if !ok {
+			sink.Spec = v1beta2.PingSourceSpec{
+				SourceSpec: source.Spec.SourceSpec,
+				Schedule:   source.Spec.Schedule,
+				Timezone:   source.Spec.Timezone,
+			}
+
+			if source.Spec.JsonData != "" {
+				msg, err := makeMessage(source.Spec.JsonData)
+				if err != nil {
+					return fmt.Errorf("error converting jsonData to a higher version: %v", err)
+				}
+				sink.Spec.ContentType = cloudevents.ApplicationJSON
+				sink.Spec.Data = string(msg)
+			}
+		}
+
+		// marshal and store v1beta1.PingSource.Spec into V1B1SpecAnnotationKey
+		// this is to help if we need to convert back to v1beta1.PingSource
+		v1beta1Spec, _ := json.Marshal(source.Spec)
```

ok to swallow the error?
maybe log it?